File: explicit-java/src/main/kotlin/explicit/token/Regex.kt
package explicit.token

import explicit.api.IToken
import explicit.api.Navigator
import explicit.evaluator.ExplicitRuleEvaluation
import explicit.parser.ExplicitRules
import java.util.*
import java.util.regex.Pattern

/**
 * Token implementation which represents a regex.
 */
data class Regex(val pattern: String): IToken {
    override fun toString() = "`$pattern`"
    override fun asText() = pattern

    override fun evaluate(rulez: ExplicitRules, navigator: Navigator<String>): ExplicitRuleEvaluation {
        val pattern = Pattern.compile(asText())

        val tuples = ArrayList<Pair<Int, String>>()
        tuples.add(Pair(navigator.getIndex() - 1, navigator.getCurr()))

        val entries = HashMap<String, String>()

        var matcher = pattern!!.matcher(tuples.map { it.second }.joinToString(""))
        if (matcher.find()) {
            entries["group"] = matcher.group()
            return ExplicitRuleEvaluation(true, HashMap(), listOf(navigator.getIndex() - 1))
        }
        while (navigator.hasNext()) {
            navigator.next()
            tuples.add(Pair(navigator.getIndex() - 1, navigator.getCurr()))

            matcher = pattern.matcher(tuples.map { it.second }.joinToString(""))
            if (matcher.find()) {
                for (i in tuples.indices.reversed()) {
                    val substr = tuples.subList(i, tuples.size)
                            .map { it.second }.joinToString("")

                    matcher = pattern.matcher(substr)
                    if (matcher.find()) {
                        if (matcher.start() == 0 || substr[matcher.start() - 1] == ' ') {
                            entries["group"] = matcher.group()
                            return ExplicitRuleEvaluation(true, entries, tuples.subList(i, tuples.size)
                                    .map { it.first })
                        }
                    }
                }
            }
        }
        return ExplicitRuleEvaluation(false, entries, ArrayList())
    }
}

File: explicit-python/explicit_nlu/token/Regex.py
import logging
import re
from dataclasses import dataclass
from typing import List, Dict

from explicit_nlu.api import IToken, Navigator
from explicit_nlu.evaluator import ExplicitEvaluation
from explicit_nlu.parser import ExplicitRules


@dataclass
class RegexUnit:
    index: int
    token: str


@dataclass
class Regex(IToken):
    pattern: str

    def __str__(self):
        return "`" + self.pattern + "`"

    def __repr__(self):
        return self.__str__()

    def as_text(self):
        return self.pattern

    def evaluate(self, rulez: ExplicitRules, navigator: Navigator) -> ExplicitEvaluation:
        logging.debug("evaluate regex")

        tuples: List[RegexUnit] = [RegexUnit(navigator.get_absolute_index() - 1, navigator.get_curr())]

        pattern = re.compile(self.pattern)
        entries: Dict[str, str] = dict()

        text = "".join(map(lambda x: x.token, tuples))
        result = re.search(pattern, text)

        if result:
            entries["group"] = result.group()
            return ExplicitEvaluation(True, entries, [navigator.get_absolute_index() - 1])

        while navigator.has_next():
            navigator.__next__()
            tuples.append(RegexUnit(navigator.get_absolute_index() - 1, navigator.get_curr()))

            text = "".join(map(lambda x: x.token, tuples))
            result = re.search(pattern, text)

            if result:
                for i in reversed(range(len(tuples))):
                    substr = "".join(map(lambda x: x.token, tuples[i:]))
                    result = re.search(pattern, text)

                    if result:
                        if result.pos == 0 and substr[result.pos - 1] == ' ':
                            entries["group"] = result.group()
                            return ExplicitEvaluation(True, entries, list(map(lambda x: x.index, tuples[i:])))

        return ExplicitEvaluation(False, entries, [])
